### DISCUSSION TITLE
The search tool encountered an error: Torch not compiled with CUDA enabled

### DISCUSSION THREAD

**Discussion by hillzzz** [open-webui/open-webui#8170] [Issues]
Hi Folks, I am getting this error "**The search tool encountered an error: Torch not compiled with CUDA enabled**" and it's driving me crazy. I am using an Nvidia Jetson AGX Orin as the platform and PyTorch with CUDA is correctly installed. I am using Jetpack 6.1 and everything works. I am Running Ollama natively not through docker and that works fine I can run models up to 70Billion parameters. I am using the CUDA enabled version of Open-webui and I set use CUDA true in the docker file and launch flag --gpus all, Open-Webui is a great front end and works really well except for the above error when doing a search, it starts with something like "loading embedding 1/3" which appears very briefly and then the error "The search tool encountered an error: Torch not compiled with CUDA enabled" appears. the LLM sometimes thinks that phrase is the search result and complains about the context.
I know CUDA is correctly installed on the AGX Orin "nv_tegra_release# R36 REVISION: 4.0" Ollama works fine and uses the GPU and I have tried everything I can think of including rewriting the Dockerfile to use genuine nvidia CUDA libs and Pytorch from the official Nvidia repository but I still get this error. Can anyone help? Thanks

**Comment by hillzzz** [comment]
I tried the non CUDA version of open-Webui 5.2 and the behaviour is exactly the same as the CUDA version. When the LLM initiates a search the same error appears "loading embedding 1/3" which appears very briefly and then the error "The search tool encountered an error: Torch not compiled with CUDA enabled". I can't discover where the error is coming from :( anyone any ideas?

**Reply by RayBytes** [reply]
hey, did you ever figure out this issue? getting the same problem

**Comment by fsdnotavai** [comment]
same problem. Torch not compiled with CUDA enabled.
i ran cmd, py, checked my torch DOES indeed is working fine with cuda. TF.
freaking reddit doesn't allow me to post because I withdraw the post because they messed up with my [02] image attachment. stupid reddit junk.
![01](https://github.com/user-attachments/assets/35162b77-ecbc-441b-aeb5-783666f43841)
![02](https://github.com/user-attachments/assets/d93ad74b-2021-4c69-9191-0aaf338fdc2e)

**Comment by awesomez** [comment]
- Are any of you using Pinokio to run Open WebUI? - Do you think this is where the bug/cause might be located?

I am having this issue too. It worked fine for one search - then nada. 


UPDATE:

It looks like it's the "pip install onnxruntime" that's the issue, both in Pinokio and in others that's the culprit here.

The issue is that it's using the CPU-only version of the Open-WebUI interface, not the GPU/cuda version. 

While there is a GPU version of the onnxruntime out there for Pinokio by editing "install.js" in the git root in Pinokio:

```
module.exports = { run: [
  { method: "shell.run", params: {
      venv: "env", path: "app",
      message: [
        "uv pip install open-webui -U",
        "uv pip install onnxruntime-gpu[cuda,cudnn]",
        "uv pip install torch torchvision torchaudio --index-url https://download.pytorch.org/whl/cu121"
      ]
  } }
] }
```
This isn't quite enough to get Pinokio to use GPU mode due to the torch versions being out of date and the security issues of running arbitrary code if this is enabled as-is using the "LLM Web Search" tool found on the OpenWebUI website.


WORKAROUND:

If you're using the Pinokio version of Open WebUI, you can simply ensure CPU-only mode 'Valve' is set to ON to bypass the Torch issue (since this is only for GPU-acceleration / CUDA stuff used in the script).


GPU Speed!!!

If you want the GPU speed while using this tool (like I did), and you're on Windows (like I am), you will need to use the Docker and/or Docker Compose instructions found in the Quick Start guide on the Open WebUI documentation to get the specific docker image with the GPU/cuda-specific support this web tool requires:  https://docs.openwebui.com/getting-started/quick-start/

Use Docker Desktop to help you manage the Docker / Compose stuff.


Fully LOCAL search:

If you're using SearXNG like I am (since I want fully local search), you'll need to pull the docker image of that as well in Docker Desktop (you can just search for it at the top of the app). 

You'll now need to use the following settings in the "LLM Web Search" tool in Open WebUI (click the tiny gear icon) under Tools listing found under "Workspace" in OWUI:

![image](https://github.com/user-attachments/assets/b4356b85-7306-4299-b1d3-0f582157e888)


Fully LOCAL models:

Since both components are located in Docker Desktop, and if you want to use LM Studio (or another OpenAI-compatible app), you'll use these settings in your Models tab to attach to the internal host's (your physical machine's) resources:

![image](https://github.com/user-attachments/assets/04fab7b7-0f3c-4888-a5dd-ca6543ec542a)

That will allow you to use LM Studio to serve your models to the newly GPU-instanced "LLM Web Tool"-powered Open WebUI image you pulled from the Quick Start guide above.

Unfortunately there is no way (at least securely) to do this inside Pinokio just yet that I know of due to the pytorch versioning being too outdated to be secure.

Hopefully this guide is helpful to someone! :)

**Comment by mamei16** [comment]
Hi guys, see my comment on the issue: https://github.com/mamei16/LLM_Web_search_OWUI/issues/12#issuecomment-3558978955
